### PR TITLE
BUG: Rewrite _make_concat_multiindex

### LINKED ISSUE #20565: concat handles MultiIndex differently when row count is the same
#### Code Sample, a copy-pastable example if possible

```python
rows1 = 3
df1 = pd.DataFrame(np.random.randn(rows1, 2), columns=['A', 'B'], index=['Z1']*rows1)

rows2 = 3
df2 = pd.DataFrame(np.random.randn(rows2, 2), columns=['A', 'B'], index=['Z1']*rows2)

df = pd.concat([df1, df2], keys=['Key1', 'Key2'], names=['KEY', 'ID'])
print(df.index)

```
#### Problem description
If rows1 = rows2 as in the example, I get the following multi index from concat where 'Z1' is repeated with separate references in labels. However, if rows1 != rows2 (e.g. change rows1 to 4), I get what seems to be the more reasonable result in the expected output where 'Z1' is listed once with the same reference in labels. 

MultiIndex(levels=[['Key1', 'Key2'], **['Z1', 'Z1', 'Z1']**],
           labels=[[0, 0, 0, 1, 1, 1], **[0, 1, 2, 0, 1, 2]**],
           names=['KEY', 'ID'])

The repeating multi index seemed to cause issues later when using inner merge on left/right index. Is this concat behavior expected?

#### Expected Output
MultiIndex(levels=[['Key1', 'Key2'], **['Z1']**],
           labels=[[0, 0, 0, 1, 1, 1], **[0, 0, 0, 0, 0, 0]**],
           names=['KEY', 'ID'])

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.22.0
pytest: 3.0.7
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.25.2
numpy: 1.12.1
scipy: 1.0.1
pyarrow: None
xarray: None
IPython: 5.3.0
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.7
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 3.7.3
bs4: 4.6.0
html5lib: 0.999
sqlalchemy: 1.2.5
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by toobaz**
- [x] closes #20565
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.25.0.rst [hunk 1]
```diff
@@ -183,6 +183,7 @@ Reshaping
 
 - Bug in :func:`merge` when merging by index name would sometimes result in an incorrectly numbered index (:issue:`24212`)
 - :func:`to_records` now accepts dtypes to its `column_dtypes` parameter (:issue:`24895`)
+- Bug in :func:`concat` creating a malformed :class:`MultiIndex` when passed multiple frames indexed by identical :class:`MultiIndex`es (:issue:`20565`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: you need a space before the es I think

File: pandas/core/reshape/concat.py [hunk 2]
```diff
@@ -533,103 +529,62 @@ def _concat_indexes(indexes):
...
-            names = list(names)
+    orig = _concat_indexes(indexes)
+
+    # Simplest way to create and prepend the keys level(s):
+    keys_chunks = [([key] * len(idx)) for (key, idx) in zip(keys, indexes)]
+    keys_levs = Index([i for l in keys_chunks for i in l],
+                      tupleize_cols=True)
+    tot_df = concat([keys_levs.to_frame().reset_index(drop=True),
+                     orig.to_frame().reset_index(drop=True)], axis=1)
+    temp_names = [None] * keys_levs.nlevels + list(orig.names)
```
Comment (jreback) [inline pandas/core/reshape/concat.py]: name this empty_ames

File: pandas/core/reshape/concat.py [hunk 3]
```diff
@@ -533,103 +529,62 @@ def _concat_indexes(indexes):
...
-
-    if len(new_names) < len(new_levels):
-        new_names.extend(new_index.names)
-
-    return MultiIndex(levels=new_levels, codes=new_codes, names=new_names,
-                      verify_integrity=False)
+            # Received names for all levels
+            result.names = names
+
+    if levels is not None:
```
Comment (jreback) [inline pandas/core/reshape/concat.py]: can you add some commentary what you are doing here

**Review by jreback — CHANGES_REQUESTED** [review]
small comments. otherwise nice cleaning.

File: pandas/core/reshape/concat.py [hunk 4]
```diff
@@ -533,103 +529,62 @@ def _concat_indexes(indexes):
...
-            levels.append(categories)
-            codes_list.append(codes)
+    """
 
-        if len(names) == len(levels):
-            names = list(names)
+    orig = _concat_indexes(indexes)
+
+    # Simplest way to create and prepend the keys level(s):
+    keys_chunks = [([key] * len(idx)) for (key, idx) in zip(keys, indexes)]
```
Comment (mroeschke) [inline pandas/core/reshape/concat.py]: Maybe make this a generator comprehension instead

**Comment by jreback** [comment]
closing as stale

**Pull request closed without merging**